File: TextosJapones/src/main/java/org/jisho/textosJapones/controller/CorrecaoController.java
package org.jisho.textosJapones.controller;

import java.io.IOException;
import java.net.URL;
import java.util.ArrayList;
import java.util.List;
import java.util.ResourceBundle;

import org.jisho.textosJapones.model.entities.Vocabulario;
import org.jisho.textosJapones.model.enums.Notificacao;
import org.jisho.textosJapones.model.exceptions.ExcessaoBd;
import org.jisho.textosJapones.model.services.VocabularioServices;
import org.jisho.textosJapones.util.notification.AlertasPopup;
import org.jisho.textosJapones.util.notification.Notificacoes;

import com.jfoenix.controls.JFXButton;
import com.jfoenix.controls.JFXDialog;
import com.jfoenix.controls.JFXDialogLayout;
import com.jfoenix.controls.JFXTextField;
import com.jfoenix.controls.events.JFXDialogEvent;

import javafx.event.EventHandler;
import javafx.fxml.FXML;
import javafx.fxml.FXMLLoader;
import javafx.fxml.Initializable;
import javafx.geometry.Insets;
import javafx.scene.Node;
import javafx.scene.Parent;
import javafx.scene.control.Label;
import javafx.scene.effect.BoxBlur;
import javafx.scene.input.KeyCode;
import javafx.scene.input.KeyEvent;
import javafx.scene.layout.StackPane;
import javafx.scene.paint.Color;
import javafx.scene.robot.Robot;
import javafx.scene.text.Font;

public class CorrecaoController implements Initializable {

	final private static String STYLE_SHEET = AlertasPopup.class
			.getResource("/org/jisho/textosJapones/resources/css/Dark_Theme.css").toExternalForm();

	@FXML
	public JFXTextField txtVocabulario;

	@FXML
	public JFXTextField txtTraducao;

	public static JFXButton btnVoltar;
	public static JFXButton btnCancelar;
	public static JFXButton btnConfirmar;

	private static JFXDialog dialog;

	private VocabularioServices vocabServ;
	private Vocabulario vocabulario;
	private Robot robot = new Robot();

	private void onBtnCancelar() {
		limpar();
	}

	private void onBtnConfirmar() {
		salvar();
	}

	private CorrecaoController servico() {
		vocabServ = new VocabularioServices();
		return this;
	}

	private CorrecaoController procurar() {
		if (!txtVocabulario.getText().trim().isEmpty()) {
			if (vocabServ == null)
				servico();

			try {
				if (vocabServ.existe(txtVocabulario.getText())) {
					vocabulario = vocabServ.select(txtVocabulario.getText().trim());
					carregar();
				} else {
					txtVocabulario.setUnFocusColor(Color.RED);
					Notificacoes.notificacao(Notificacao.ERRO, "Vocabulário informado não encontrado.",
							txtVocabulario.getText());
				}

			} catch (ExcessaoBd e) {
				e.printStackTrace();
				Notificacoes.notificacao(Notificacao.ERRO, "Erro ao carregar vocabulário.", txtVocabulario.getText());
				txtVocabulario.setUnFocusColor(Color.RED);
			}
		}
		return this;
	}

	private CorrecaoController salvar() {
		if (!txtTraducao.getText().trim().isEmpty()) {
			if (vocabServ == null)
				servico();

			try {
				atualiza();
				vocabServ.insertOrUpdate(vocabulario);
				Notificacoes.notificacao(Notificacao.SUCESSO, "Vocabulário salvo com sucesso.",
						vocabulario.getTraducao());
				limpar();
				txtVocabulario.requestFocus();
			} catch (ExcessaoBd e) {
				e.printStackTrace();
				Notificacoes.notificacao(Notificacao.ERRO, "Erro ao salvar tradução.", txtTraducao.getText());
			}
		} else
			txtVocabulario.setUnFocusColor(Color.RED);

		return this;
	}

	private CorrecaoController carregar() {
		txtVocabulario.setText(vocabulario.getVocabulario());
		txtTraducao.setText(vocabulario.getTraducao());
		return this;
	}

	private CorrecaoController atualiza() {
		vocabulario.setTraducao(txtTraducao.getText().trim());
		return this;
	}

	private CorrecaoController limpar() {
		txtVocabulario.setText("");
		txtTraducao.setText("");
		vocabulario = null;
		return this;
	}

	public static void abreTelaCorrecao(StackPane rootStackPane, Node nodeBlur) {
		try {
			BoxBlur blur = new BoxBlur(3, 3, 3);
			JFXDialogLayout dialogLayout = new JFXDialogLayout();
			dialog = new JFXDialog(rootStackPane, dialogLayout, JFXDialog.DialogTransition.CENTER);

			FXMLLoader loader = new FXMLLoader();
			loader.setLocation(getFxmlLocate());
			Parent newAnchorPane = loader.load();
			CorrecaoController cnt = loader.getController();

			Label titulo = new Label("Tela de correção");
			titulo.setFont(Font.font(20));
			titulo.setTextFill(Color.web("#ffffff", 0.8));

			List<JFXButton> botoes = new ArrayList<JFXButton>();

			btnConfirmar = new JFXButton("Confirmar");
			btnConfirmar.setOnAction(AE -> cnt.onBtnConfirmar());
			btnConfirmar.getStyleClass().add("background-Green2");
			botoes.add(btnConfirmar);

			btnCancelar = new JFXButton("Cancelar");
			btnCancelar.setOnAction(AC -> cnt.onBtnCancelar());
			btnCancelar.getStyleClass().add("background-Red2");
			botoes.add(btnCancelar);

			btnVoltar = new JFXButton("Voltar");
			btnVoltar.setOnAction(AV -> dialog.close());
			btnVoltar.getStyleClass().add("background-White1");
			botoes.add(btnVoltar);

			dialogLayout.setHeading(titulo);
			dialogLayout.setBody(newAnchorPane);
			dialogLayout.setActions(botoes);

			dialog.getStylesheets().add(STYLE_SHEET);
			dialog.setPadding(new Insets(0, 0, 0, 0));
			dialog.setOnDialogClosed((JFXDialogEvent event1) -> {
				nodeBlur.setEffect(null);
			});

			nodeBlur.setEffect(blur);
			dialog.show();
		} catch (IOException e) {

			e.printStackTrace();
		}
	}

	private void configuraListenert() {
		txtVocabulario.focusedProperty().addListener((o, oldVal, newVal) -> {
			if (oldVal) {
				txtVocabulario.setUnFocusColor(Color.web("#106ebe"));
				procurar();
			}
		});

		txtTraducao.focusedProperty().addListener((o, oldVal, newVal) -> {
			if (oldVal)
				txtTraducao.setUnFocusColor(Color.web("#106ebe"));
		});

		txtVocabulario.setOnKeyPressed(new EventHandler<KeyEvent>() {
			@Override
			public void handle(KeyEvent ke) {
				if (ke.getCode().equals(KeyCode.ENTER))
					robot.keyPress(KeyCode.TAB);
			}
		});

		txtTraducao.setOnKeyPressed(new EventHandler<KeyEvent>() {
			@Override
			public void handle(KeyEvent ke) {
				if (ke.getCode().equals(KeyCode.ENTER))
					robot.keyPress(KeyCode.TAB);
			}
		});
	}

	@Override
	public void initialize(URL location, ResourceBundle resources) {
		configuraListenert();
		servico();
	}

	public static URL getFxmlLocate() {
		return CorrecaoController.class.getResource("/org/jisho/textosJapones/view/Correcao.fxml");
	}
}

File: ProcessaTexto/src/main/kotlin/br/com/fenix/processatexto/controller/CorrecaoController.kt
package br.com.fenix.processatexto.controller

import br.com.fenix.processatexto.components.notification.AlertasPopup
import br.com.fenix.processatexto.components.notification.Notificacoes
import br.com.fenix.processatexto.model.entities.processatexto.Vocabulario
import br.com.fenix.processatexto.model.enums.Notificacao
import br.com.fenix.processatexto.service.VocabularioJaponesServices
import com.jfoenix.controls.JFXButton
import com.jfoenix.controls.JFXDialog
import com.jfoenix.controls.JFXDialogLayout
import com.jfoenix.controls.JFXTextField
import com.jfoenix.controls.events.JFXDialogEvent
import javafx.fxml.FXML
import javafx.fxml.FXMLLoader
import javafx.fxml.Initializable
import javafx.geometry.Insets
import javafx.scene.Node
import javafx.scene.Parent
import javafx.scene.control.Label
import javafx.scene.effect.BoxBlur
import javafx.scene.input.KeyCode
import javafx.scene.layout.StackPane
import javafx.scene.paint.Color
import javafx.scene.robot.Robot
import javafx.scene.text.Font
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.io.IOException
import java.net.URL
import java.sql.SQLException
import java.util.*


class CorrecaoController : Initializable {

    @FXML
    lateinit var txtVocabulario: JFXTextField

    @FXML
    lateinit var txtTraducao: JFXTextField

    private var vocabServ: VocabularioJaponesServices? = null
    private var vocabulario: Optional<Vocabulario> = Optional.empty()
    private val robot: Robot = Robot()

    private fun onBtnCancelar() = limpar()

    private fun onBtnConfirmar() = salvar()

    private fun servico(): CorrecaoController {
        vocabServ = VocabularioJaponesServices()
        return this
    }

    private fun procurar(): CorrecaoController {
        if (txtVocabulario.text.trim().isNotEmpty()) {

            if (vocabServ == null)
                servico()

            try {
                if (vocabServ!!.existe(txtVocabulario.text)) {
                    vocabulario = vocabServ!!.select(txtVocabulario.text.trim())
                    carregar()
                } else {
                    txtVocabulario.unFocusColor = Color.RED
                    Notificacoes.notificacao(Notificacao.ERRO, "Vocabulário informado não encontrado.", txtVocabulario.text)
                }
            } catch (e: SQLException) {
                LOGGER.error(e.message, e)
                Notificacoes.notificacao(Notificacao.ERRO, "Erro ao carregar vocabulário.", txtVocabulario.text)
                txtVocabulario.unFocusColor = Color.RED
            }
        }
        return this
    }

    private fun salvar(): CorrecaoController {
        if (txtTraducao.text.trim().isNotEmpty()) {
            if (vocabServ == null)
                servico()

            try {
                atualiza()
                vocabServ!!.insertOrUpdate(vocabulario.get())
                Notificacoes.notificacao(Notificacao.SUCESSO, "Vocabulário salvo com sucesso.", vocabulario.get().portugues)
                limpar()
                txtVocabulario.requestFocus()
            } catch (e: SQLException) {
                LOGGER.error(e.message, e)
                Notificacoes.notificacao(Notificacao.ERRO, "Erro ao salvar tradução.", txtTraducao.text)
            }
        } else txtVocabulario.unFocusColor = Color.RED
        return this
    }

    private fun carregar(): CorrecaoController {
        txtVocabulario.text = vocabulario.get().vocabulario
        txtTraducao.text = vocabulario.get().portugues
        return this
    }

    private fun atualiza(): CorrecaoController {
        vocabulario.get().portugues = txtTraducao.text.trim()
        return this
    }

    private fun limpar(): CorrecaoController {
        txtVocabulario.text = ""
        txtTraducao.text = ""
        vocabulario = Optional.empty()
        return this
    }

    private fun configuraListenert() {
        txtVocabulario.focusedProperty().addListener { _, oldVal, _ ->
            if (oldVal) {
                txtVocabulario.unFocusColor = Color.web("#106ebe")
                procurar()
            }
        }
        txtTraducao.focusedProperty().addListener { _, oldVal, _ -> if (oldVal) txtTraducao.unFocusColor = Color.web("#106ebe") }
        txtVocabulario.setOnKeyPressed { ke -> if (ke.code.equals(KeyCode.ENTER)) robot.keyPress(KeyCode.TAB) }
        txtTraducao.setOnKeyPressed { ke -> if (ke.code.equals(KeyCode.ENTER)) robot.keyPress(KeyCode.TAB) }
    }

    @Override
    override fun initialize(location: URL, resources: ResourceBundle) {
        configuraListenert()
        servico()
    }

    companion object {
        private val LOGGER: Logger = LoggerFactory.getLogger(CorrecaoController::class.java)

        private val STYLE_SHEET: String = AlertasPopup::class.java.getResource("/css/Dark_Theme.css").toExternalForm()

        fun abreTelaCorrecao(rootStackPane: StackPane, nodeBlur: Node) {
            try {
                val blur = BoxBlur(3.0, 3.0, 3)
                val dialogLayout = JFXDialogLayout()

                val dialog = JFXDialog(rootStackPane, dialogLayout, JFXDialog.DialogTransition.CENTER)
                val loader = FXMLLoader()
                loader.location = fxmlLocate
                val newAnchorPane: Parent = loader.load()
                val cnt: CorrecaoController = loader.getController()
                val titulo = Label("Tela de correção")
                titulo.font = Font.font(20.0)
                titulo.textFill = Color.web("#ffffff", 0.8)

                val botoes: MutableList<JFXButton> = mutableListOf()
                val btnConfirmar = JFXButton("Confirmar")
                btnConfirmar.setOnAction { cnt.onBtnConfirmar() }
                btnConfirmar.styleClass.add("background-Green2")
                botoes.add(btnConfirmar)
                val btnCancelar = JFXButton("Cancelar")
                btnCancelar.setOnAction { cnt.onBtnCancelar() }
                btnCancelar.styleClass.add("background-Red2")
                botoes.add(btnCancelar)
                val btnVoltar = JFXButton("Voltar")
                btnVoltar.setOnAction { dialog.close() }
                btnVoltar.styleClass.add("background-White1")
                botoes.add(btnVoltar)
                dialogLayout.setHeading(titulo)
                dialogLayout.setBody(newAnchorPane)
                dialogLayout.setActions(botoes)

                dialog.stylesheets.add(STYLE_SHEET)
                dialog.padding = Insets(0.0, 0.0, 0.0, 0.0)
                dialog.setOnDialogClosed { nodeBlur.effect = null }
                nodeBlur.effect = blur
                dialog.show()
            } catch (e: IOException) {
                LOGGER.error(e.message, e)
            }
        }

        val fxmlLocate: URL get() = CorrecaoController::class.java.getResource("/view/Correcao.fxml")
    }
}